### ISSUE TITLE
LN SWAP

### ISSUE THREAD

**Issue by kilrau** [ExchangeUnion/xud#348]
No more words needed. Should be your sole focus for `alpha.1` (aka `Berlin` release) @sangaman 

![image](https://user-images.githubusercontent.com/19181985/44173825-ece80500-a095-11e8-85b6-64273978e2c8.png)

**Comment by offerm** [comment]
What needs to be done:
1. P2P communication between the taker and maker. Starting with finding a match. To reduce risk we will also implement the swap staring by `executeSwap` cli.  Initially, we will skip the LND work and use stubs for that. This allows preparing for a demo while integrating to LND. @offerm
2. RPC - implementing the `swap-resolver` service. This service is used by LND to resolve a `r-hash` into `pre-image` @sangaman 
3. Integration with LNDs for initiate a swap and for resolving the hash - @offerm
3. Order book integration. @sangaman 

For now, we focus mainly on the `happy flow`. We deal with error situations later or one the time pressure reduced.

It is important to work on a detailed demo script (are we doing a demo in Berlin??).  It would be bad to make changes at the last minute due to changes to the demo. @kilrau 

In parallel we need to prepare test1,test2, test3 for LN swap (install BTCD, LTCD, LNDx2). Also, decide which machine we will do the demo and make sure it is ready for it. @kilrau @sangaman 

Create a branch for that. https://github.com/ExchangeUnion/xud/tree/LN-Swap-Berlin
@sangaman I suggest that changes you/others made to it will be in a PR

**Comment by sangaman** [comment]
Are you thinking of having the `executeSwap` command necessary for triggering a swap, aka the SWAP_REQUEST packet? If so, why not just have it be automatic after the DEAL_RESPONSE packet? I don't see the risk.

I'll probably need some guidance with the swap-resolver service, I'll reach out to you.

Are you suggesting we make PRs against the LN-Swap-Berlin branch rather than master? I think I'd want to stick to feature branch PRs against master because I'd be afraid of major merge conflicts between master and LN-Swap-Berlin after two weeks.

And yes, only happy path at this point but let's try to make notes in issues and TODO comments on what we'll need to do to handle errors.

**Comment by sangaman** [comment]
`DealRequestPacket` - Taker to maker. Specifies the details of the swap like what's being swapped and how much. Currently explicitly defines all that data but will be changed to point to an actual order.
`DealResponsePacket` - Maker to taker. Accepts or rejects the deal (currently always accepts) and send the `r_hash`.
`SwapRequestPacket` - Taker to maker. Basically reaffirms that the maker should continue with the swap agreed to in the first two packets. Probably we can manage without this eventually, having the maker begin the swap as soon the deal is accepted.
`SwapResponsePacket` - Maker to taker. After the swap is finished, the maker sends the `r_preimage` to the taker. We could also probably do without this eventually as well, as the taker's lnd will know the preimage at this point.

**Comment by offerm** [comment]
So, I think I finished my task for this. Both swap-resolver and executeSwap of XUD are working good for me.
The challenge now is to make it work for you @sangaman and @kilrau before we dive into checking the orderBook. So lets do it:

Step 1: Make sure you are up to date with your LND, swap-resolver, XUD:
- LND: git@github.com:offerm/lnd.git branch resolver. Must have commit `3ee6b86abecc6d6850e7c98bcd417bf68dea6ef7`. Make sure to run `make && make install` and restart all your LND processes (4 of them) 
- swap-resolver: git@github.com:offerm/swap-resolver.git branch master. Must have commit `6ce04c9576c27f45d3fdf5bdef0e8d3f54d8866c`. Make sure to restart the swap-resolver process (2 of them)
- XUD : branch LN-swap-berlin (unless merge to master). Make sure to compile.

Step 2: update aliases and include --network testnet in them.
```
alias xa-lnd-btc='lncli --network testnet --rpcserver=localhost:10002 --no-macaroons'
alias xa-lnd-ltc='lncli --network testnet --rpcserver=localhost:10001 --no-macaroons'
alias xb-lnd-btc='lncli --network testnet --rpcserver=localhost:20002 --no-macaroons'
alias xb-lnd-ltc='lncli --network testnet --rpcserver=localhost:20001 --no-macaroons'
```
Make sure you source the file

Step 3: start the swap-resolver processes using ./start_swap_resolver. The should listen on 8885 and 8895. If you have XUD processes working on these ports. please stop them.

Step 4: connect the nodes and execure a swap to make sure it is working:
- cd $GOPATH/src/github/offerm/swap-resolver
- go run cmd/xucli/main.go --rpcserver localhost:8886 takeorder --order_id=123 --maker_amount 200 --maker_coin BTC --taker_amount 10000 --taker_coin=LTC

Make sure you get back the preimage
```
2018/08/25 16:16:07 Starting takeOrder command -  (*swapresolver.TakeOrderReq)(0xc4200a3c80)(orderid:"123" taker_amount:10000 taker_coin:LTC maker_amount:200 )
2018/08/25 16:16:07 Swap completed successfully.
  Swap preImage is  4324947c70dd324e0e149e888a0842869b2519c6d97d1120489162c2399d823c 
```

This indicates that all is well

Step 5: stop the swap-resolver processe and start xud process using ./start.bash. Note that you need to define `XUDHOME` before that. set it to when your XUD project is.  The XUD proceses are using exchangeA and exchnageB databases.  Make sure they started and connected to the LNDs properly. No setup needs to be done by you. If it is not working, let me know.

Step 5:
- connects the XUD nodes.
- execure swap with `./bin/xucli  -p 8896  executeSwap takeR 1 btc 111 ltc <pubkey>`

You should see "success" on screen. check the taker's XUD log and make sure you see this:
`debug: [SWAP] Swap completed. preimage = 1d7142d4d74be1f1ce0cd4a6d832cd448deb1d8e038ce37afb3555aff3e8c5a5`


Now you can start with orderBook....

If something is not working, plaese let me know (gitter).

**Comment by sangaman** [comment]
Works for me!

```
C:\Users\Daniel\go\src\github.com\lightningnetwork\lnd>lncli --no-macaroons --rpcserver=localhost:10002 listchannels
{
    "channels": [
        {
            "active": true,
            "remote_pubkey": "02c0959690c7858cae84709324f3d7f6799d865c61451be0153141da165cb32b36",
            "channel_point": "faf80f76e3e968a3cdd1772c8a8cd826c72fad7f2859670e46008163455c2ee4:0",
            "chan_id": "1549603309848559616",
            "capacity": "10000000",
            "local_balance": "4790263",
            "remote_balance": "5100000",
            "commit_fee": "109737",
            "commit_weight": "724",
            "fee_per_kw": "151571",
            "unsettled_balance": "0",
            "total_satoshis_sent": "100000",
            "total_satoshis_received": "0",
            "num_updates": "9",
            "pending_htlcs": [
            ],
            "csv_delay": 1201,
            "private": false
        }
    ]
}

C:\Users\Daniel\go\src\github.com\lightningnetwork\lnd>lncli --no-macaroons --rpcserver=localhost:10001 listchannels
{
    "channels": [
        {
            "active": true,
            "remote_pubkey": "03aeee74cdd489936a4db75bf3865ad24e38c1c22911705bb49529e335a23b3212",
            "channel_point": "7fa61db57f46c252a31251c88d16c0fdb92eca841df0e4b257ecb16ad2beb23e:0",
            "chan_id": "784269549464846336",
            "capacity": "100000000",
            "local_balance": "59995475",
            "remote_balance": "40000000",
            "commit_fee": "4525",
            "commit_weight": "724",
            "fee_per_kw": "6250",
            "unsettled_balance": "0",
            "total_satoshis_sent": "0",
            "total_satoshis_received": "10000000",
            "num_updates": "2",
            "pending_htlcs": [
            ],
            "csv_delay": 801,
            "private": false
        }
    ]
}
```

See the satoshis received and satoshis sent.

**Comment by offerm** [comment]
@sangaman safe to close this one?

**Comment by sangaman** [comment]
It will be closed when #399 is merged.
